File: graphs/802-eventual-safe-states/solution.clj
(ns solution)

(defn is-safe? [graph node safe]
  (if (contains? safe node)
    (safe node)
    (let [safe (assoc safe node false) ; Mark as unsafe initially
          adj-nodes (graph node)]
      (if (some (fn [adj] (not (is-safe? graph adj safe))) adj-nodes)
        false
        (assoc safe node true)))))

(defn eventual-safe-nodes [graph]
  (reduce (fn [ans i]
            (if (is-safe? graph i {}) (conj ans i) ans)) [] (range (count graph))))

(def graph [[1, 2] [2, 3] [5] [0] [5] [] []])

(eventual-safe-nodes graph)

File: graphs/802-eventual-safe-states/solution.js
/**
 * @param {number[][]} graph
 * @return {number[]}
 */
var eventualSafeNodes = function (graph) {
    // Brute: try to find a terminal node for each node!
    const safe = {}

    // Start by iterating over the array, explore for each path;
    let ans = [];
    for (let i = 0; i < graph.length; i++) {
        if (isSafe(graph, i)) ans.push(i)
    }
    return ans;

    // Recursively check if the children are safe
    function isSafe(graph, node) {
        if (node in safe) return safe[node];

        // Negative Space
        safe[node] = false;
        if (graph[node].some(adj => !isSafe(graph, adj))) return false;

        // Positive Space
        safe[node] = true;
        return true;
    }
};
